### ISSUE TITLE
Advice to hosters with Engintron.

### ISSUE THREAD

**Issue by SashRaven** [engintron/engintron#314]
TL;DR use your own nameservers as resolvers.

Hi!
I'm an system administrator with  a couple cPanel servers used for shared hosting.
I were interested in using Engintron on my servers, but one thing was stopping me - cloudflare and other proxy what users can use.
Sure, Engintron have custom_rules for this case, but this is not an exit for server with hundreds clients and thousands sites, because client may at any time switch DNS to CL or some proxy server and hide IP.
Sure, I may ask client's to contact me before doing so, so I'll able to add domain to custom_rules. But, again - this is a bad idea and not very comfortable for end-users and doing an additional job for me.
So, I sit and began to think - how I can get real IP in any case, even is domain have hidden IP.
Real IP, in any case is stored... on my nameservers! If domain exist on my cpanel - it's dns exist on my DNS servers, not depending on what NS domain is pointed to. 
So, here is a solution in nginx.conf:

resolver    ip1 ip2 ip3 valid=600s ipv6=off;

Where:
ip1 - external IP of localhost (mostly it's eth0). We have an named on local, so why not to use it and reduce networks lookups?
ip2 - ip of the first nameserver
ip3 - ip of the second nameserver

valid=600s - caching time for dns answers. By default nginx using dns ttl. You can override it.
 ipv6=off - I don't use ipv6, so turn off.

With this setup I don't bother at all about getting real IP for Engintron. 
Second step - we must allow nameservers for recursive lookups through:
On NS open named.conf and add such sections with your hosting servers IP:

acl "trusted" {
        1.1.1.1;
        2.2.2.2;
        3.3.3.3;
 };

Find and edit bellow:

allow-recursion { trusted; };

Restart named.
I'm using such setup on 5 shared hosting server for 3 months and have no issues.
Hope I help someone with this.

**Comment by fevangelou** [comment]
This is kind of a hackish solution if I may say. The problem with CloudFlare is that it hides the domain's original IP. That's why you need to specify the IPs for these domains. However if you have a single shared IP, you essentially specify this IP and you're done, with or without CloudFlare. The only reason to specify domain/IP combinations is when you have both shared and dedicated IPs.

**Comment by SashRaven** [comment]
This is not a hackish at all. I'm only using my own resolvers. And this resolve "The problem with CloudFlare is that it hides the domain's original IP". 
I have thousands of sites with shared and dedicated IPs. And I don't  bother anymore if some of my customer will transfer domain to CloudFlare at any time.

**Comment by fevangelou** [comment]
It's hackish in the sense that not all servers have a DNS server installed and setup internally. If you do have local DNS resolvers, it will sure work. I'll see if there's a way to detect active local DNS and if so, append the defined nameserver IPs as resolvers in Nginx's configuration.

**Comment by fevangelou** [comment]
In the meantime I'll add your solution to the wiki.
